storage/storage_test.go

```diff
@@ -1344,3 +1360,149 @@ func TestWithEndpoint(t *testing.T) {
+		originalScheme := c.scheme
+
+		op := "bucket creation"
+
+		c.Bucket("test-bucket").Create(ctx, "pid", nil)
+		if err != nil {
+			t.Errorf("%s: %v", tc.desc, err)
+		}
+
+		if c.raw.BasePath != originalRawBasePath {
+			t.Errorf("%s: raw.BasePath changed after %s\n\tgot %v, original %v", tc.desc, op, c.raw.BasePath, originalRawBasePath)
+		}
+		if c.readHost != originalReadHost {
+			t.Errorf("%s: readHost changed after %s\n\tgot %v, original %v", tc.desc, op, c.readHost, originalReadHost)
+		}
+		if c.scheme != originalScheme {
+			t.Errorf("%s: scheme changed after %s\n\tgot %v, original %v", tc.desc, op, c.scheme, originalScheme)
+		}
+
+		op = "object upload"
+
+		w := c.Bucket("test-bucket").Object("file").NewWriter(ctx)
+		_, err = io.Copy(w, strings.NewReader("copyng into bucket"))
+		if err != nil {
+			t.Errorf("%s: %v", tc.desc, err)
+		}
+		if closeErr := w.Close(); closeErr != nil {
+			t.Errorf("%s: %v", tc.desc, closeErr)
+		}
+
+		if c.raw.BasePath != originalRawBasePath {
+			t.Errorf("%s: raw.BasePath changed after %s\n\tgot %v, original %v", tc.desc, op, c.raw.BasePath, originalRawBasePath)
+		}
+		if c.readHost != originalReadHost {
+			t.Errorf("%s: readHost changed after %s\n\tgot %v, original %v", tc.desc, op, c.readHost, originalReadHost)
+		}
+		if c.scheme != originalScheme {
+			t.Errorf("%s: scheme changed after %s\n\tgot %v, original %v", tc.desc, op, c.scheme, originalScheme)
+		}
+
+		op = "object download"
+
+		rc, err := c.Bucket("test-bucket").Object("file").NewReader(ctx)
+		if err != nil {
+			t.Errorf("%s: %v", tc.desc, err)
+		}
+		defer rc.Close()
+
+		_, err = io.Copy(ioutil.Discard, rc)
+		if err != nil {
+			t.Errorf("%s: %v", tc.desc, err)
+		}
+
+		if c.raw.BasePath != originalRawBasePath {
+			t.Errorf("%s: raw.BasePath changed after %s\n\tgot %v, original %v", tc.desc, op, c.raw.BasePath, originalRawBasePath)
+		}
+		if c.readHost != originalReadHost {
+			t.Errorf("%s: readHost changed after %s\n\tgot %v, original %v", tc.desc, op, c.readHost, originalReadHost)
+		}
+		if c.scheme != originalScheme {
+			t.Errorf("%s: scheme changed after %s\n\tgot %v, original %v", tc.desc, op, c.scheme, originalScheme)
+		}
+
+		op = "bucket delete"
+
+		if err := c.Bucket("test-bucket").Delete(ctx); err != nil {
+			t.Errorf("%s: %v", tc.desc, err)
+		}
+
+		if c.raw.BasePath != originalRawBasePath {
+			t.Errorf("%s: raw.BasePath changed after %s\n\tgot %v, original %v", tc.desc, op, c.raw.BasePath, originalRawBasePath)
+		}
+		if c.readHost != originalReadHost {
+			t.Errorf("%s: readHost changed after %s\n\tgot %v, original %v", tc.desc, op, c.readHost, originalReadHost)
+		}
+		if c.scheme != originalScheme {
+			t.Errorf("%s: scheme changed after %s\n\tgot %v, original %v", tc.desc, op, c.scheme, originalScheme)
+		}
+	}
+}
+
+func newTestServerAt(handler func(w http.ResponseWriter, r *http.Request)) (string, *http.Client, func()) {
+	ts := httptest.NewTLSServer(http.HandlerFunc(handler))
```

Comment:
Can you have the test server verify that the expected endpoint is being hit? That is probably more important than validating that the client fields are not being changed.

Reply:
I added some code for checking that the same endpoint is hit after completing the operations on the client, is that what you had in mind?

Reply:
No, I'd like to have the server verify that the calls made to it are as expected (for example create bucket resulted in a call to `POST https://<hostname>/storage/v1/b` and not to some other endpoint). This will ensure that something problematic isn't happening at the apiary layer for example.